README.md
# AlarmClock
PiZero Alarm Clock



Changelog.txt
Changelog

1 - setup gpio output ports
2 - Added functions to convert time to time in seconds
3 - added GPIO interupt for IR inputs
4 - Added pickle load to AlarmClock_Alarm.py


Temp_to_LCD.py
from time import sleep, strftime
from datetime import datetime
import Adafruit_CharLCD as LCD
import Adafruit_MCP9808.MCP9808 as MCP9808
import CHIP_IO.GPIO as GPIO

#intialise LCD
lcd = LCD.Adafruit_CharLCDPlate()

#set initial background colour
lcd.set_color(1.0,0.0,1.0) # blue LCD only
lcd.set_backlight(0)

#intialise temp sensor
sensor = MCP9808.MCP9808()
sensor.begin()

#Function to get temperature
def Get_Temperature():
    """
    Get_Tepearture Function:
    Function to get the temperature in degrees C from the attached sensor

    Args: None

    Return: Temperature (as string)
    """
    try:
        temp = sensor.readTempC()
        temp = str('%.0f' % temp) # convert to string of 0 decimal places
        return temp
    except KeyboardInterrupt:
        print ('KeyboardInterrupt in Function Get_Temperature')
    except:
        print ('Error in Function Get_Temperature')
    finally:
        GPIO.cleanup()

#fucntion to get system time
def Get_ActualTime():
    ActualTimeDisplay = datetime.now().strftime('%H:%M')
    Hour = int(datetime.now().strftime('%H'))
    Min = int(datetime.now().strftime('%M'))
    Sec = int(datetime.now().strftime('%S'))
    ActualTime = {'H':Hour, 'M':Min, 'S':Sec}
    return ActualTime, ActualTimeDisplay

def Get_Date():
    Date = datetime.now().strftime('%a %d %b')
    return Date

def F_BackLightON():
    lcd.set_backlight(1)
    sleep(10)
    lcd.set_backlight(0)

#main
while True:
    if lcd.is_pressed(LCD.SELECT):
        F_BackLightON()

    ActualTime, ActualTimeDisplay = Get_ActualTime()
    ActualTime_in_seconds = ((ActualTime['H']*60*60) + (ActualTime['M']*60) + ActualTime['S'])
    date = Get_Date()
    temp = Get_Temperature()
    lcd.set_cursor(0,0)
    lcd.message(ActualTimeDisplay)
    lcd.set_cursor(12,0)
    lcd.message(temp + chr(223) + 'C') #chr(223) is degree sign
    lcd.set_cursor(3,1)
    lcd.message(date)
    sleep(0.5)


AlarmClock_Alarm.py
# encoding: utf-8
"""
AlarmClock_Alarm.py

Script 2 of 2 from AlarmClock Project:
Script for comparing alarm time to actual time, triggering the alarm and disabling the alarm

Created by Christopher Beard on 16-07-2016.
Copyright (c) 2016 notice:
This code is shared under the Creative Commons Attribution-ShareAlike
4.0 International Public License
It is also shared under the GNU GENERAL PUBLIC LICENSE Version 3
"""

from time import sleep
import AlarmClock_Screen
import pickle
import RPi.GPIO as GPIO
GPIO.setup(port, GPIO.OUT, [initial=0]) #set ports later
GPIO.setup(port, GPIO.OUT, [initial=0])
GPIO.setup(port, GPIO.IN, [pull_up_down=GPIO.PUD_DOWN])
GPIO.setup(port, GPIO.IN, [pull_up_down=GPIO.PUD_DOWN])


def Get_ActualTime_in_Seconds():
    """
    Get_ActualTime_in_Seconds Function:
    Function gets actual time from AlarmClock_Screen Script and then converts to seconds

    Args: None

    Returns: ActualTime_in_seconds
    """
    try:
        ActualTime, ActualTimeDisplay = AlarmClock_Screen.Get_ActualTime()  # get as dictionary object Hour,Mins,Secs
        ActualTime_in_seconds = ((ActualTime['H']*60*60) + (ActualTime['M']*60) + ActualTime['S'])
        return ActualTime_in_seconds
    except KeyboardInterrupt:
        print ('KeyboardInterrupt in Function Get_ActualTime_in_Seconds')
    finally:
        GPIO.cleanup()

def Get_AlarmTime_in_Seconds(Alarm_Number):
    """
    Get_AlarmTime_in_Seconds Function:
    Function gets alarm time from AlarmClock_Screen Script and then converts to seconds

    Args: Alarm_Number

    Returns: AlarmTime_in_seconds
    """
    try:
        #use pickle to get data from other script
        if Alarm_Number == 1:
            with open('AlarmTime1.pickle', 'rb') as f:
            #get as dictionary object
            AlarmTime = pickle.load(f)
        elif Alarm_Number == 2:
            with open('AlarmTime2.pickle', 'rb') as f:
            AlarmTime = pickle.load(f)
        else:
            print ('AlarmNumber passed to Function Get_AlarmTime_in_Seconds not valid ')

        AlarmTime_in_seconds = ((AlarmTime['H']*60*60) + (AlarmTime['M']*60) + AlarmTime['S'])
        return AlarmTime_in_seconds
    except KeyboardInterrupt:
        print ('KeyboardInterrupt in Function Get_AlarmTime_in_Seconds')
    finally:
        GPIO.cleanup()

def GPIO_Call(Buzzer, LEDFlash):
    """
    GPIO_Call Function:
    Take in Booleans and sets corresponding GPIO Outputs

    Args:Buzzer, LEDFlash

    Returns: None
    """
    try:
        if Buzzer:
            GPIO.output(port, 1) #set port later
        else:
            GPIO.output(port, 0)

        if LEDFlash:
            GPIO.output(port, 1)
        else:
            GPIO.output(port, 0)
    except KeyboardInterrupt:
        print ('KeyboardInterrupt in Function GPIO_Call')
    finally:
        GPIO.cleanup()

def F_Alarm_Active(AlarmTime, ActualTime, AlarmButtonPressed, LEDFlash, Buzzer, AlarmHappened):
    """
    F_Alarm_Active Function:
    Takes in Alarm parameters the triggers alarm at correct time

    Args: AlarmTime, ActualTime, AlarmButtonPressed, LEDFlash, Buzzer, AlarmHappened

    Returns: AlarmHappened, AlarmButtonPressed, LEDFlash, Buzzer
    """
    try:
        # if the alarm has gone off and then the user deactivates it sleep for 20 mins
        if AlarmButtonPressed and AlarmHappened:
            Buzzer = False
            LEDFlash = False
            AlarmHappened = False
            AlarmButtonPressed = False
            GPIO_Call(Buzzer, LEDFlash)
            sleep(20*60) #20 mins in seconds
        # If actual time is greater than the alarm time + 10 mins from above e.g. 03:00 > 02:40 then
        # the alarm can never go off as 24hr clock so actual time will cycle round at midnight
        elif ActualTime >= (AlarmTime + (10*60)):
            AlarmHappened = False
            Buzzer = False
            LEDFlash = False
            GPIO_Call(Buzzer, LEDFlash)
            sleep(0.25) # preserve processor
        # Set condition for if alarm has gone off and user hasnt deactivated after 10 secs of beeping and 2 min silence
        elif (ActualTime >= (AlarmTime + (2*60))) and (ActualTime < (AlarmTime + (10*60))):
            if Buzzer:
                Buzzer = False
                LEDFlash = False
            elif not Buzzer:
                Buzzer = True
                LEDFlash = True
            GPIO_Call(Buzzer, LEDFlash)
            sleep(1)
        # set condition for 30 secs before alarm time
        elif (ActualTime >= (AlarmTime - 30)) and (ActualTime < (AlarmTime - 10)):
            AlarmHappened = True
            if LEDFlash:
                LEDFlash = False
            elif not LEDFlash:
                LEDFlash = True
            Buzzer = False
            GPIO_Call(Buzzer, LEDFlash)
            sleep(1)
        #set condition for 10 secs before alarm time (beeping)
        elif (ActualTime >= (AlarmTime - 10)) and (ActualTime < AlarmTime):
            AlarmHappened = True
            if Buzzer:
                Buzzer = False
                LEDFlash = False
            elif not Buzzer:
                Buzzer = True
                LEDFlash = True
            GPIO_Call(Buzzer, LEDFlash)
            sleep(1)
        # set condition for 2 min slience period of alarm
        elif (ActualTime >= AlarmTime) and (ActualTime < (AlarmTime + (2*60))):
            AlarmHappened = True
            Buzzer = False
            LEDFlash = False
            GPIO_Call(Buzzer, LEDFlash)
            sleep(0.25) # preserve processor
        # set condition for sleeping alarm if over 10 mins to go
        elif (ActualTime < (AlarmTime - (10*60))):
            sleep(0.25) # preserve processor

        return [AlarmHappened, AlarmButtonPressed, LEDFlash, Buzzer]

    except KeyboardInterrupt:
        print ("Keyboard Interruption in F_Alarm_Active")
    finally:
        GPIO.cleanup()

def F_Alarm1ButtonPressed(Alarm1ButtonPressed):
    return Alarm1ButtonPressed = True

def F_Alarm2ButtonPressed(Alarm2ButtonPressed):
    return Alarm2ButtonPressed = True

def main():
    """
    Main Funtion:
    Runs in constant while loop checking status of alarm (on/off),
    If alarm on it calls F_Alarm_Active function
    """
    #set up initial variables
    Alarm1_Happened = False
    Alarm2_Happened = False
    Alarm1ButtonPressed = False # IR Button Input for deactivating alarm
    Alarm2ButtonPressed = False
    LEDFlash = False
    Buzzer = False
    Alarm_Number = None

    try:
        while True:
            with open('Alarm1Active.pickle' , 'rb') as A1:
                Alarm1Active = pickle.load(A1)

            if Alarm1Active:
                Alarm_Number = 1
                Alarm1Time = Get_AlarmTime_in_Seconds(Alarm_Number)
                ActualTime = Get_ActualTime_in_Seconds()
                GPIO.add_event_detect(port, GPIO.RISING, callback=F_Alarm1ButtonPressed(Alarm1ButtonPressed), bouncetime=300)
                [Alarm1_Happened, Alarm1ButtonPressed, LEDFlash, Buzzer] = F_Alarm_Active(Alarm1Time, ActualTime, \
                    Alarm1ButtonPressed, LEDFlash, Buzzer, Alarm1_Happened)
            else:
                sleep(0.5) #preserve processor

            with open('Alarm2Active.pickle' , 'rb') as A2:
                Alarm1Active = pickle.load(A2)

            if Alarm2Active:
                Alarm_Number = 2
                Alarm2Time = Get_AlarmTime_in_Seconds(Alarm_Number)
                ActualTime = Get_ActualTime_in_Seconds()
                GPIO.add_event_detect(port, GPIO.RISING, callback=F_Alarm2ButtonPressed(Alarm2ButtonPressed), bouncetime=300)
                [Alarm2_Happened, Alarm2ButtonPressed, LEDFlash, Buzzer] = F_Alarm_Active(Alarm2Time, ActualTime, \
                    Alarm2ButtonPressed, LEDFlash, Buzzer, Alarm2_Happened)
            else:
                sleep(0.5) #preserve processor

    except KeyboardInterrupt:
        print ("Keyboard Interuption in main")
    finally:
        GPIO.cleanup()


if __name__ == '__main__':
    main()


AlarmClock_Alarm_Debug.py
# encoding: utf-8
"""
AlarmClock_Alarm.py

Script 2 of 2 from AlarmClock Project:
Script for comparing alarm time to actual time, triggering the alarm and disabling the alarm

Created by Christopher Beard on 16-07-2016.
#Copyright statement, share but attribute
"""

from time import sleep
import AlarmClock_Screen_Debug
#import RPi
import pickle

def Get_ActualTime_in_Seconds():
    """
    Get_ActualTime_in_Seconds Function:
    Function gets actual time from AlarmClock_Screen Script and then converts to seconds

    Args: None

    Returns: ActualTime_in_seconds
    """
    try:
        ActualTime, ActualTimeDisplay = AlarmClock_Screen_Debug.Get_ActualTime()  # get as dictionary object Hour,Mins,Secs
        ActualTime_in_seconds = ((ActualTime['H']*60*60) + (ActualTime['M']*60) + ActualTime['S'])
        return ActualTime_in_seconds
    except KeyboardInterrupt:
        print ('KeyboardInterrupt in Function Get_ActualTime_in_Seconds')


def GPIO_Call(Buzzer, LEDFlash):
    """
    GPIO_Call Fucntion:
    Take in Booleans and sets corresponding GPIO Outputs

    Args:Buzzer, LEDFlash

    Returns: None
    """
    # set up GPIO
    if Buzzer == True:
        pass # set GPIO true
    else:
        pass # set GPIO false

    if LEDFlash == True:
        pass # set GPIO True
    else:
        pass # set GPIO False


def Alarm_Active(AlarmTime, ActualTime, AlarmButtonPressed, LEDFlash, Buzzer, AlarmHappened):
    """
    Alarm_Active Function:
    Takes in Alarm parameters the triggers alarm at correct time

    Args: AlarmTime, ActualTime, AlarmButtonPressed, LEDFlash, Buzzer, AlarmHappened

    Returns: AlarmHappened, AlarmButtonPressed, LEDFlash, Buzzer
    """

    try:
        # if the alarm has gone off and then the user deactivates it sleep for 20 mins
        if (AlarmButtonPressed == True) and (AlarmHappened == True):
            print ('1')
            Buzzer = False
            LEDFlash = False
            AlarmHappened = False
            AlarmButtonPressed = False
            GPIO_Call(Buzzer, LEDFlash)
            #sleep(20*60) #20 mins in seconds
            sleep(10)
        # If actual time is greater than the alarm time + 10 mins from above e.g. 03:00 > 02:40 then
        # the alarm can never go off as 24hr clock so actual time will cycle round at midnight
        elif ActualTime >= (AlarmTime + (10*60)):
            print ('2')
            AlarmHappened = False
            Buzzer = False
            LEDFlash = False
            GPIO_Call(Buzzer, LEDFlash)
            sleep(0.25) # sleep for two hours, reduce for initial debug, maybe inly sleep for 0.25 to reduce processor
        # Set condition for if alarm has gone off and user hasnt deactivated after 10 secs of beeping and 2 min silence
        elif (ActualTime >= (AlarmTime + (2*60))) and (ActualTime < (AlarmTime + (10*60))):
            print ('3')
            if Buzzer == True:
                Buzzer = False
                LEDFlash = False
            elif Buzzer == False:
                Buzzer = True
                LEDFlash = True
            GPIO_Call(Buzzer, LEDFlash)
            sleep(1)
        # set condition for 30 secs before alarm time
        elif (ActualTime >= (AlarmTime - 30)) and (ActualTime < (AlarmTime - 10)):
            print ('4')
            AlarmHappened = True
            if LEDFlash == True:
                LEDFlash = False
            elif LEDFlash == False:
                LEDFlash = True
            Buzzer = False
            GPIO_Call(Buzzer, LEDFlash)
            print (Buzzer)
            print (LEDFlash)
            sleep(1)
        #set condition for 10 secs before alarm time (beeping)
        elif (ActualTime >= (AlarmTime - 10)) and (ActualTime < AlarmTime):
            print ('5')
            AlarmHappened = True
            if Buzzer == True:
                Buzzer = False
                LEDFlash = False
            elif Buzzer == False:
                Buzzer = True
                LEDFlash = True
            GPIO_Call(Buzzer, LEDFlash)
            print (Buzzer)
            print (LEDFlash)
            sleep(1)
        # set condition for 2 min slience period of alarm
        elif (ActualTime >= AlarmTime) and (ActualTime < (AlarmTime + (2*60))):
            print ('6')
            AlarmHappened = True
            Buzzer = False
            LEDFlash = False
            GPIO_Call(Buzzer, LEDFlash)
            sleep(0.25)
        # set condition for sleeping alarm if over 30 secsto go
        elif (ActualTime < (AlarmTime - (30))):
            print ('7')
            sleep(0.25)
        else:
            print ('8')

        return [AlarmHappened, AlarmButtonPressed, LEDFlash, Buzzer]

    except KeyboardInterrupt:
        #GPIO_Cleanup()
        print ("Keyboard Interruption in Alarm_Active")


def main():
    """
    Main Funtion:
    Runs in constant while loop checking status of alarm (on/off),
    If alarm on it calls Alarm_Active function
    """
    #set up initial variables
    Alarm1Time = (17*60*60)
    Alarm1_Happened = False  # Can you do arrays python or does list work?
    #Alarm2_Happened = False
    Alarm1ButtonPressed = False # IR Button Input for deactivating alarm
    #Alarm2ButtonPressed = False
    LEDFlash = False
    Buzzer = False
    #Alarm_Number = None

    try:
        while True:
            with open('Alarm1Active.pickle' , 'rb') as A1:
                Alarm1Active = pickle.load(A1)

            if Alarm1Active == True:
                #Alarm_Number = 1
                #Alarm1Time = AlarmClock_Screen.Get_AlarmTime(Alarm_Number)
                ActualTime = Get_ActualTime_in_Seconds()
                #AlarmButtonGPIO = None # Assign Later
                #set up input GPIO as interupt if high then set Alarm1ButtonPressed = True
                #call function
                print ('Going into function')
                [Alarm1_Happened, Alarm1ButtonPressed, LEDFlash, Buzzer] = Alarm_Active(Alarm1Time, ActualTime, \
                    Alarm1ButtonPressed, LEDFlash, Buzzer, Alarm1_Happened) # check syntax
            else:
                sleep(1) #sleep for 1 hour in seconds to preserve battery, take out for initial debug
                print('Alarm Not Active')

            #if AlarmClock_Screen.Alarm2Active == True:
             #   Alarm_Number = 2
              #  Alarm2Time = AlarmClock_Screen.Get_AlarmTime(Alarm_Number)
               # ActualTime = AlarmClock_Screen.Get_ActualTime()
                #AlarmButtonGPIO = None # Assign Later
                #set up input GPIO as interupt if high then set Alarm1ButtonPressed = True
                #call function
                #Alarm2_Happened, Alarm2ButtonPressed, LEDFlash, Buzzer = Alarm_Active(Alarm2Time, ActualTime, \
               #     Alarm2ButtonPressed, LEDFlash, Buzzer, Alarm2_Happened) # check syntax
            #else:
             #   sleep(3600) #sleep for 1 hour in seconds to preserve battery, take out for initial debug and pass

    except KeyboardInterrupt:
        #GPIO_Cleanup() # check correct
        print ("Keyboard Interuption in main")

if __name__ == '__main__':
    main()


AlarmClock_Screen_Debug.py
# encoding: utf-8
"""
AlarmClock_screen.py

Script 1 of 2 from AlarmClock Project:
Script for controlling the LCD Screen on the alarm clock, displaying time and date,
reading the set alarm time etc.

Created by Christopher Beard on 16-07-2016.
Copyright (c) 2016 notice:
This code is shared under the Creative Commons Attribution-ShareAlike
4.0 International Public License
It is also shared under the GNU GENERAL PUBLIC LICENSE Version 3
"""
from time import sleep, strftime
from datetime import datetime
import pickle
import Adafruit_CharLCD as LCD
import Adafruit_MCP9808.MCP9808 as MCP9808
import CHIP_IO.GPIO as GPIO
# GPIO.setmode(GPIO.BCM)
# GPIO.setup(port, GPIO.IN, [pull_up_down=GPIO.PUD_DOWN])
#intialise LCD
lcd = LCD.Adafruit_CharLCDPlate()
#set initial background colour
lcd.set_color(1.0,0.0,1.0) # blue LCD only
#set background off intially, need GPIO event to set high
lcd.set_backlight(0)
#intialise temp sensor
sensor = MCP9808.MCP9808()
sensor.begin()



def Get_Temperature():
    """
    Get_Tepearture Function:
    Function to get the temperature in degrees C from the attached sensor

    Args: None

    Return: Temperature (as string)
    """
    try:
        temp = sensor.readTempC()
        temp = str('%.0f' % temp) # convert to string of 0 decimal places
        return temp
    except KeyboardInterrupt:
        print ('KeyboardInterrupt in Function Get_Temperature')
    finally:
        GPIO.cleanup()

def Get_ActualTime():
    """
    Get_ActualTime Function:
    Function to get the actual time from the system, then seperate into
    a time for the display and a dictionary item for the time in
    individual hours, minutes and seconds

    Args: None

    Return: ActualTime,  ActualTimeDisplay
    """
    try:
        ActualTimeDisplay = datetime.now().strftime('%H:%M')
        Hour = int(datetime.now().strftime('%H'))
        Min = int(datetime.now().strftime('%M'))
        Sec = int(datetime.now().strftime('%S'))
        ActualTime = {'H':Hour, 'M':Min, 'S':Sec}
        return ActualTime, ActualTimeDisplay
    except KeyboardInterrupt:
        print ('KeyboardInterrupt in Function Get_ActualTime')
    finally:
        GPIO.cleanup()

def Get_Date():
    """
    Get_Date fucntion:
    Fucntion to get date from the system

    Args: None

    Return: Date
    """
    try:
        Date = datetime.now().strftime('%a %d %b')
        return Date
    except KeyboardInterrupt:
        print ('KeyboardInterrupt in Function Get_Date')
    finally:
        GPIO.cleanup()

def Set_AlarmTime(Hour, Min, Sec, Alarm_Number):
    lcd.set_backlight(1)
    lcd.clear()
    lcd.message('Alarm ' + str(Alarm_Number) + ' Time:\n' )
    lcd.message(str(Hour)+':'+str(Min))
    Cycle = True
    while Cycle:
        if lcd.is_pressed(LCD.UP):
            Hour += 1
            lcd.clear()
            lcd.message('Alarm ' + str(Alarm_Number) + ' Time:\n' )
            lcd.message(str(Hour)+':'+str(Min))
        elif lcd.is_pressed(LCD.DOWN):
            Hour -= 1
            lcd.clear()
            lcd.message('Alarm ' + str(Alarm_Number) + ' Time:\n' )
            lcd.message(str(Hour)+':'+str(Min))
        elif lcd.is_pressed(LCD.SELECT):
            Cycle = False
            sleep(0.5)

    Cycle = True
    while Cycle:
        if lcd.is_pressed(LCD.UP):
            Min += 1
            lcd.clear()
            lcd.message('Alarm ' + str(Alarm_Number) + ' Time:\n' )
            lcd.message(str(Hour)+':'+str(Min))
        elif lcd.is_pressed(LCD.DOWN):
            Min -= 1
            lcd.clear()
            lcd.message('Alarm ' + str(Alarm_Number) + ' Time:\n' )
            lcd.message(str(Hour)+':'+str(Min))
        elif lcd.is_pressed(LCD.SELECT):
            Cycle = False
            sleep(0.5)

    lcd.clear()
    lcd.set_backlight(0)
    return Hour, Min, Sec

def Get_AlarmTime(Hour, Min, Sec, Alarm_Number):
    #convert to dictionary
    AlarmTime = {'H':Hour, 'M':Min, 'S':Sec}
    return AlarmTime

def F_BackLightON():
    """
    F_BackLightON function:
    Function to turn LCD backlight on for a duration of time

    Args: None

    Return: None
    """
    try:
        lcd.set_backlight(1)
        sleep(8)
        lcd.set_backlight(0)
    except KeyboardInterrupt:
        print ('KeyboardInterrupt in Function F_BackLightON')
    finally:
        GPIO.cleanup()


def F_AlarmOnOff(AlarmActive, Alarm_Number):
    """
    AlarmActive function:
    Function to enable and disable alarm when button is pressed as well as
    display status on screen

    Args: AlarmActive

    Return: AlarmActive
    """
    try:
        if AlarmActive:
            AlarmActive = False
            lcd.set_backlight(1)
            lcd.clear()
            lcd.message('Alarm ' + str(Alarm_Number) + ' Disabled')
            sleep(2)
            lcd.clear()
            lcd.set_backlight(0)
            return AlarmActive
        else:
            AlarmActive = True
            lcd.set_backlight(1)
            lcd.clear()
            lcd.message('Alarm ' + str(Alarm_Number) + ' Enabled')
            sleep(2)
            lcd.clear()
            lcd.set_backlight(0)
            return AlarmActive
    except KeyboardInterrupt:
        print ('KeyboardInterrupt in Function F_BackLightON')
    finally:
        GPIO.cleanup()

def main():
    """
    Main Function:
    Runs continously calling appropriate functions to set screens depending on buttons pressed
    """
    try:
        #Intialise variables
        Alarm1Active = False
        Alarm2Active = False
        Hour1 = 1
        Min1 = 1
        Sec1 = 0

        while True:
            if lcd.is_pressed(LCD.SELECT):
                F_BackLightON()
            elif lcd.is_pressed(LCD.UP):
                Alarm1Active = F_AlarmOnOff(Alarm1Active, 1)
            elif lcd.is_pressed(LCD.DOWN):
                Alarm2Active = F_AlarmOnOff(Alarm2Active, 2)
            elif lcd.is_pressed(LCD.LEFT):
                Hour1, Min1, Sec1 = Set_AlarmTime(Hour1, Min1, Sec1, 1)
            else:
                #Main Screen
                ActualTime, ActualTimeDisplay = Get_ActualTime()
                date = Get_Date()
                temp = Get_Temperature()
                lcd.set_cursor(0,0)
                lcd.message(ActualTimeDisplay)
                lcd.set_cursor(12,0)
                lcd.message(temp + chr(223) + 'C') #chr(223) is degree sign
                lcd.set_cursor(3,1)
                lcd.message(date)
                sleep(0.25)


            # #trigger event for backlight
            # GPIO.add_event_detect(port, GPIO.RISING, callback=F_BackLightON(), bouncetime=300)
            # #tigger event for enabling/disabling first alarm
            # GPIO.add_event_detect(port, GPIO.RISING, callback=F_AlarmOnOff(Alarm1Active), bouncetime=300)
            # #setup pickles here
            with open('Alarm1Active.pickle', 'wb') as f:
                # Pickle the 'data' using the highest protocol available.
                pickle.dump(Alarm1Active, f, pickle.HIGHEST_PROTOCOL)
            with open('AlarmTime1.pickle', 'wb') as f3:
                AlarmTime1 = Get_AlarmTime(Hour1, Min1, Sec1, 1)
                print (AlarmTime1)
                # Pickle the 'data' using the highest protocol available.
                pickle.dump(AlarmTime1, f3, pickle.HIGHEST_PROTOCOL)
    except KeyboardInterrupt:
        print('KeyboardInterrupt in Main')
    finally:
        GPIO.cleanup()

if __name__ == '__main__':
    main()


AlarmClock_Screen.py
# encoding: utf-8
"""
AlarmClock_screen.py

Script 1 of 2 from AlarmClock Project:
Script for controlling the LCD Screen on the alarm clock, displaying time and date,
reading the set alarm time etc.

Created by Christopher Beard on 16-07-2016.
Copyright (c) 2016 notice:
This code is shared under the Creative Commons Attribution-ShareAlike
4.0 International Public License
It is also shared under the GNU GENERAL PUBLIC LICENSE Version 3
"""
from time import sleep, strftime
from datetime import datetime
import pickle
import Adafruit_CharLCD as LCD
import Adafruit_MCP9808.MCP9808 as MCP9808
import RPi.GPIO as GPIO
GPIO.setmode(GPIO.BCM)
GPIO.setup(port, GPIO.IN, [pull_up_down=GPIO.PUD_DOWN])
#intialise LCD
lcd = LCD.Adafruit_CharLCDPlate()
#set initial background colour
lcd.set_color(1.0,0.0,1.0) # blue LCD only
#set background off intially, need GPIO event to set high
lcd.set_backlight(0)
#intialise temp sensor
sensor = MCP9808.MCP9808()
sensor.begin()

def Get_Temperature():
    """
    Get_Tepearture Function:
    Function to get the temperature in degrees C from the attached sensor

    Args: None

    Return: Temperature (as string)
    """
    try:
        temp = sensor.readTempC()
        temp = str('%.0f' % temp) # convert to string of 0 decimal places
        return temp
    except KeyboardInterrupt:
        print ('KeyboardInterrupt in Function Get_Temperature')
    finally:
        GPIO.cleanup()

def Get_ActualTime():
    """
    Get_ActualTime Function:
    Function to get the actual time from the system, then seperate into
    a time for the display and a dictionary item for the time in
    individual hours, minutes and seconds

    Args: None

    Return: ActualTime,  ActualTimeDisplay
    """
    try:
        ActualTimeDisplay = datetime.now().strftime('%H:%M')
        Hour = int(datetime.now().strftime('%H'))
        Min = int(datetime.now().strftime('%M'))
        Sec = int(datetime.now().strftime('%S'))
        ActualTime = {'H':Hour, 'M':Min, 'S':Sec}
        return ActualTime, ActualTimeDisplay
    except KeyboardInterrupt:
        print ('KeyboardInterrupt in Function Get_ActualTime')
    finally:
        GPIO.cleanup()

def Get_Date():
    """
    Get_Date fucntion:
    Fucntion to get date from the system

    Args: None

    Return: Date
    """
    try:
        Date = datetime.now().strftime('%a %d %b')
        return Date
    except KeyboardInterrupt:
        print ('KeyboardInterrupt in Function Get_Date')
    finally:
        GPIO.cleanup()

def Set_AlarmTime(Hour, Min, Sec, Alarm_Number):
    """
    Set_AlarmTime Function:
    Function to alter the set alarm time using the LCD touchbuttons changing
    hour and minute seperately

    Args: Hour, Min, Sec, Alarm_Number

    Return: Hour, Min, Sec
    """
    lcd.set_backlight(1)
    lcd.clear()
    lcd.message('Alarm ' + str(Alarm_Number) + ' Time:\n' )
    lcd.message(str(Hour)+':'+str(Min))
    Cycle = True
    while Cycle:
        if lcd.is_pressed(LCD.UP):
            Hour += 1
            lcd.clear()
            lcd.message('Alarm ' + str(Alarm_Number) + ' Time:\n' )
            lcd.message(str(Hour)+':'+str(Min))
        elif lcd.is_pressed(LCD.DOWN):
            Hour -= 1
            lcd.clear()
            lcd.message('Alarm ' + str(Alarm_Number) + ' Time:\n' )
            lcd.message(str(Hour)+':'+str(Min))
        elif lcd.is_pressed(LCD.SELECT):
            Cycle = False
            sleep(0.5)

    Cycle = True
    while Cycle:
        if lcd.is_pressed(LCD.UP):
            Min += 1
            lcd.clear()
            lcd.message('Alarm ' + str(Alarm_Number) + ' Time:\n' )
            lcd.message(str(Hour)+':'+str(Min))
        elif lcd.is_pressed(LCD.DOWN):
            Min -= 1
            lcd.clear()
            lcd.message('Alarm ' + str(Alarm_Number) + ' Time:\n' )
            lcd.message(str(Hour)+':'+str(Min))
        elif lcd.is_pressed(LCD.SELECT):
            Cycle = False
            sleep(0.5)

    lcd.clear()
    lcd.set_backlight(0)
    return Hour, Min, Sec

def Get_AlarmTime(Hour, Min, Sec, Alarm_Number):
    """
    Get_AlarmTime function:
    Function to get the individual times of the alarm in Hous, mins and secs
    and convert to dictionary item

    Args: Hour, Min, Sec, Alarm_Number

    Return: AlarmTime
    """
    try:
        #convert to dictionary
        AlarmTime = {'H':Hour, 'M':Min, 'S':Sec}
        return AlarmTime
    except KeyboardInterrupt:
        print ('KeyboardInterrupt in Function Get_AlarmTime')
    finally:
        GPIO.cleanup()


def F_BackLightON():
    """
    F_BackLightON function:
    Function to turn LCD backlight on for a duration of time

    Args: None

    Return: None
    """
    try:
        lcd.set_backlight(1)
        sleep(8)
        lcd.set_backlight(0)
    except KeyboardInterrupt:
        print ('KeyboardInterrupt in Function F_BackLightON')
    finally:
        GPIO.cleanup()


def F_AlarmOnOff(AlarmActive, Alarm_Number):
    """
    AlarmActive function:
    Function to enable and disable alarm when button is pressed as well as
    display status on screen

    Args: AlarmActive

    Return: AlarmActive
    """
    try:
        if AlarmActive:
            AlarmActive = False
            lcd.set_backlight(1)
            lcd.clear()
            lcd.message('Alarm ' + str(Alarm_Number) + ' Disabled')
            sleep(2)
            lcd.clear()
            lcd.set_backlight(0)
            return AlarmActive
        else:
            AlarmActive = True
            lcd.set_backlight(1)
            lcd.clear()
            lcd.message('Alarm ' + str(Alarm_Number) + ' Enabled')
            sleep(2)
            lcd.clear()
            lcd.set_backlight(0)
            return AlarmActive
    except KeyboardInterrupt:
        print ('KeyboardInterrupt in Function F_AlarmOnOff')
    finally:
        GPIO.cleanup()

def main():
    """
    Main Function:
    Runs continously calling appropriate functions to set screens depending on buttons pressed
    """
    try:
        #Intialise variables
        Alarm1Active = False
        Alarm2Active = False
        Hour1 = 1
        Min1 = 1
        Sec1 = 0
        Hour2 = 1
        Min2 = 1
        Sec2 = 0

        while True:
            if lcd.is_pressed(LCD.UP):
                #trigger event to enable/disable alarm 1
                Alarm1Active = F_AlarmOnOff(Alarm1Active, 1)
            elif lcd.is_pressed(LCD.DOWN):
                #trigger event to enable/disable alarm 2
                Alarm2Active = F_AlarmOnOff(Alarm2Active, 2)
            elif lcd.is_pressed(LCD.LEFT):
                #trigger event to set alarm 1 time
                Hour1, Min1, Sec1 = Set_AlarmTime(Hour1, Min1, Sec1, 1)
            elif lcd.is_pressed(LCD.RIGHT):
                #trigger event to set alarm 2 time
                Hour2, Min2, Sec2 = Set_AlarmTime(Hour2, Min2, Sec2, 2)
            else:
                #Main Screen
                ActualTime, ActualTimeDisplay = Get_ActualTime()
                date = Get_Date()
                temp = Get_Temperature()
                lcd.set_cursor(0,0)
                lcd.message(ActualTimeDisplay)
                lcd.set_cursor(12,0)
                lcd.message(temp + chr(223) + 'C') #chr(223) is degree sign
                lcd.set_cursor(3,1)
                lcd.message(date)
                sleep(0.25)
                #trigger event for backlight
                GPIO.add_event_detect(port, GPIO.RISING, callback=F_BackLightON(), bouncetime=300)

                #setup pickles here to pass data to AlarmClock_Alarm.py
                with open('Alarm1Active.pickle', 'wb') as f:
                    # Pickle the 'data' using the highest protocol available.
                    pickle.dump(Alarm1Active, f, pickle.HIGHEST_PROTOCOL)
                with open('Alarm2Active.pickle', 'wb') as f2:
                    # Pickle the 'data' using the highest protocol available.
                    pickle.dump(Alarm2Active, f2, pickle.HIGHEST_PROTOCOL)
                with open('AlarmTime1.pickle.', 'wb') as f3:
                    AlarmTime1 = Get_AlarmTime(Hour1, Min1, Sec1, 1)
                    # Pickle the 'data' using the highest protocol available.
                    pickle.dump(AlarmTime1, f3, pickle.HIGHEST_PROTOCOL)
                with open('AlarmTime2.pickle.', 'wb') as f4:
                    AlarmTime2 = Get_AlarmTime(Hour2, Min2, Sec2, 2)
                    # Pickle the 'data' using the highest protocol available.
                    pickle.dump(AlarmTime2, f4, pickle.HIGHEST_PROTOCOL)

    except KeyboardInterrupt:
        print('KeyboardInterrupt in Main')
    finally:
        GPIO.cleanup()

if __name__ == '__main__':
    main()
